File: smartadapter/src/main/java/smartadapter/internal/mapper/ViewHolderMapper.java
package smartadapter.internal.mapper;

/*
 * Created by Manne Öhlund on 30/05/17.
 * Copyright © 2019 All rights reserved.
 */

import android.util.SparseArray;
import android.view.ViewGroup;

import java.lang.reflect.Constructor;
import java.util.HashMap;

import smartadapter.SmartRecyclerAdapter;
import smartadapter.internal.utils.ReflectionUtils;
import smartadapter.listener.ViewEventListener;
import smartadapter.viewholder.SmartAdapterHolder;
import smartadapter.viewholder.SmartViewHolder;
import smartadapter.viewholder.ViewEventHolder;
import smartadapter.widget.ViewTypeResolver;

/**
 * Responsible for mapping subtype of SmartViewHolder with target data type.
 */
public class ViewHolderMapper {

    private final Object caller;

    private int identifier = 0;
    private final SparseArray<Class<? extends SmartViewHolder>> viewTypeMapper = new SparseArray<>();
    private final ViewHolderConstructorMapper viewHolderConstructorMapper = new ViewHolderConstructorMapper();
    private HashMap<String, Class<? extends SmartViewHolder>> dataTypeViewHolderMapper = new HashMap<>();
    private HashMap<Class<? extends SmartViewHolder>, SmartRecyclerAdapter> smartRecyclerAdapterMapper = new HashMap<>();

    public ViewHolderMapper(Object callerEnclosingClass) {
        this.caller = callerEnclosingClass;
    }

    /**
     * Will first check if viewTypeResolver is assigned and contains the viewHolder.
     * If viewHolder was found, proceed with view id mapping.
     * - Add the id if not existing
     * - Return mapping key
     *
     * @param viewTypeResolver custom view holder type resolver
     * @param item target item
     * @param position adapter position
     * @throws RuntimeException if mapping/relation of ViewHolder to data item was not found
     * @return target view holder type identifier
     */
    public int getItemViewType(ViewTypeResolver viewTypeResolver, Object item, int position) {
        Class<? extends SmartViewHolder> viewHolder = viewTypeResolver == null ? null : viewTypeResolver.getViewType(item, position);
        if (viewHolder == null) {
            viewHolder = dataTypeViewHolderMapper.get(item.getClass().getName());
        }

        if (viewHolder != null) {
            if (viewTypeMapper.indexOfValue(viewHolder) < 0) {
                viewTypeMapper.put(identifier++, viewHolder);
            }
            Class<? extends SmartViewHolder> viewType;
            for (int i = 0; i < viewTypeMapper.size(); i++) {
                viewType = viewTypeMapper.get(i);
                if (viewType.getName().equals(viewHolder.getName())) {
                    return viewTypeMapper.keyAt(i);
                }
            }
        }

        throw new RuntimeException(String.format("Fatal error! Mapping of ViewHolder to item '%s' does not exist", item.getClass().getName()));
    }

    /**
     * Instantiates the target view holder and set view event listeners if specified.
     *
     * @param viewEventListeners Contains general purpose or view id specific event listeners.
     * @param parent Parent ViewGroup
     * @param viewType View holder type
     * @param <VH> Subtype of SmartViewHolder
     * @return Target view holder
     */
    @SuppressWarnings("unchecked")
    public <VH extends SmartViewHolder> VH createViewHolder(
            HashMap<Class<? extends SmartViewHolder>, HashMap<Integer, HashMap<Integer, ViewEventListener>>> viewEventListeners,
            ViewGroup parent,
            int viewType) {
        VH viewHolder;
        Class<? extends SmartViewHolder> smartViewHolderClass = viewTypeMapper.get(viewType);
        Constructor constructor = viewHolderConstructorMapper.getConstructor(smartViewHolderClass);

        try {
            viewHolder = (VH) ReflectionUtils.invokeConstructor(constructor,
                    ReflectionUtils.isNonStaticInnerClass(smartViewHolderClass)
                            ? new Object[]{caller, parent}
                            : new Object[]{parent}
            );
        } catch (Exception e) {
            throw new RuntimeException(String.format("Could not invoke constructor for '%s', '%s'", smartViewHolderClass.toString(), e.getMessage()), e);
        }

        /*
         * Set view event listeners.
         * First check if viewEventListeners has general SmartViewHolder event listener.
         * Then if viewEventListeners has viewHolder.getClass() event listener, will override smartListener.
         */
        if (viewHolder instanceof ViewEventHolder) {
            ((ViewEventHolder)viewHolder).setViewEventListeners(viewEventListeners.get(SmartViewHolder.class));
            ((ViewEventHolder)viewHolder).setViewEventListeners(viewEventListeners.get(viewHolder.getClass()));
        }

        SmartRecyclerAdapter smartRecyclerAdapter = smartRecyclerAdapterMapper.get(viewHolder.getClass());
        if (viewHolder instanceof SmartAdapterHolder && smartRecyclerAdapter != null) {
            ((SmartAdapterHolder)viewHolder).setSmartRecyclerAdapter(smartRecyclerAdapter);
        }

        return viewHolder;
    }

    public void addMapping(Class<?> itemType, Class<? extends SmartViewHolder> viewHolderType) {
        dataTypeViewHolderMapper.put(itemType.getName(), viewHolderType);
        viewHolderConstructorMapper.add(viewHolderType);
    }

    public void setDataTypeViewHolderMapper(HashMap<String, Class<? extends SmartViewHolder>> dataTypeViewHolderMapper) {
        this.dataTypeViewHolderMapper = dataTypeViewHolderMapper;
        viewHolderConstructorMapper.add(dataTypeViewHolderMapper.values());
    }

    public void setSmartRecyclerAdapterMapper(HashMap<Class<? extends SmartViewHolder>, SmartRecyclerAdapter> smartRecyclerAdapterMapper) {
        this.smartRecyclerAdapterMapper = smartRecyclerAdapterMapper;
    }
}

File: smartadapter/src/main/java/smartadapter/internal/mapper/ViewHolderMapper.kt
package smartadapter.internal.mapper

/*
 * Created by Manne Öhlund on 30/05/17.
 * Copyright © 2019 All rights reserved.
 */

import android.util.SparseArray
import android.view.ViewGroup
import smartadapter.SmartRecyclerAdapter
import smartadapter.internal.utils.ReflectionUtils
import smartadapter.viewholder.SmartAdapterHolder
import smartadapter.viewholder.SmartViewHolder
import smartadapter.widget.ViewTypeResolver
import java.util.*
import kotlin.reflect.KClass
import kotlin.reflect.KFunction

/**
 * Responsible for mapping subtype of SmartViewHolder with target data type.
 */
class ViewHolderMapper {

    private var smartViewHolderClass: KClass<out SmartViewHolder<*>>? = null
    private var constructor: KFunction<Any>? = null
    private val viewTypeMapper = SparseArray<KClass<out SmartViewHolder<*>>>()
    private val viewHolderConstructorMapper = ViewHolderConstructorMapper()
    private var dataTypeViewHolderMapper = HashMap<String, KClass<out SmartViewHolder<*>>>()
    private var smartRecyclerAdapterMapper = HashMap<KClass<out SmartViewHolder<*>>, SmartRecyclerAdapter>()

    /**
     * Will first check if viewTypeResolver is assigned and contains the smartViewHolderClass.
     * If smartViewHolderClass was found, proceed with view id mapping.
     * - Add the id if not existing
     * - Return mapping smartViewHolderKey
     *
     * @param viewTypeResolver custom view holder type resolver
     * @param item target item
     * @param position adapter position
     * @throws RuntimeException if mapping/relation of ViewHolder to data item was not found
     * @return target view holder type identifier
     */
    fun getItemViewType(viewTypeResolver: ViewTypeResolver?, item: Any, position: Int): Int {
        smartViewHolderClass = viewTypeResolver?.getViewType(item, position)
        if (smartViewHolderClass == null) {
            smartViewHolderClass = dataTypeViewHolderMapper[item.javaClass.name]
        } else {
            viewHolderConstructorMapper.add(smartViewHolderClass!!)
        }
        if (smartViewHolderClass != null) {
            return viewTypeMapper.put(smartViewHolderClass!!)
        }

        throw RuntimeException(String.format("Fatal error! Mapping of ViewHolder to item '%s' does not exist", item.javaClass.name))
    }

    /**
     * Instantiates the target view holder and set view event listeners if specified.
     *
     * @param parent Parent ViewGroup
     * @param viewType View holder type
     * @param <VH> Subtype of SmartViewHolder
     * @return Target view holder
     */
    @Suppress("UNCHECKED_CAST")
    fun <VH : SmartViewHolder<*>> createViewHolder(
            parent: ViewGroup,
            viewType: Int): VH {
        val viewHolder: VH
        smartViewHolderClass = viewTypeMapper.get(viewType)
        constructor = viewHolderConstructorMapper.getConstructor(smartViewHolderClass!!)

        try {
            viewHolder = ReflectionUtils.invokeConstructor(constructor!!,
                    *when(ReflectionUtils.isInnerClass(smartViewHolderClass!!)) {
                        true -> arrayOf<Any?>(null, parent)
                        false -> arrayOf<Any?>(parent)
                    }
            ) as VH
        } catch (e: Exception) {
            throw RuntimeException(String.format("Could not invoke constructor for '%s', '%s'", smartViewHolderClass!!.toString(), e.message), e)
        }

        val smartRecyclerAdapter = smartRecyclerAdapterMapper[viewHolder::class]
        if (viewHolder is SmartAdapterHolder && smartRecyclerAdapter != null) {
            (viewHolder as SmartAdapterHolder).setSmartRecyclerAdapter(smartRecyclerAdapter)
        }

        return viewHolder
    }

    fun addMapping(itemType: KClass<*>, viewHolderType: KClass<out SmartViewHolder<*>>) {
        dataTypeViewHolderMapper[itemType.java.name] = viewHolderType
        viewHolderConstructorMapper.add(viewHolderType)
    }

    fun setDataTypeViewHolderMapper(dataTypeViewHolderMapper: HashMap<String, KClass<out SmartViewHolder<*>>>) {
        this.dataTypeViewHolderMapper = dataTypeViewHolderMapper
        viewHolderConstructorMapper.add(dataTypeViewHolderMapper.values)
    }

    fun setSmartRecyclerAdapterMapper(smartRecyclerAdapterMapper: HashMap<KClass<out SmartViewHolder<*>>, SmartRecyclerAdapter>) {
        this.smartRecyclerAdapterMapper = smartRecyclerAdapterMapper
    }
}

fun <E : KClass<out SmartViewHolder<*>>> SparseArray<E>.put(value: E): Int {
    val key = value.java.name.hashCode()
    put(key, value)
    return key
}
